File: client/src/lib/hearItTiming.ts
/**
 * Hear-it timing helpers — map lyric lines into Deezer preview / Spotify seek.
 * Pure (no DOM) so it can be unit-tested.
 *
 * Strategy: center the clip on the *word* (not just the LRC line start) so
 * Hear-it plays the sung word in sync with the highlighted phrase — language-agnostic.
 */

/** Cap sparse LRC gaps so word estimates stay near the stamped line start. */
const MAX_LINE_DUR_FOR_WORD_EST_SEC = 5;

export function lineBoundsSec(input: {
  timestamp_ms: number;
  line_end_ms?: number | null;
}): { lineStartSec: number; lineEndSec: number; lineDurSec: number } {
  const lineStartSec = Math.max(0, Number(input.timestamp_ms) / 1000 || 0);
  const rawEnd = input.line_end_ms != null ? Number(input.line_end_ms) / 1000 : NaN;
  const lineEndSec =
    Number.isFinite(rawEnd) && rawEnd > lineStartSec + 0.35
      ? rawEnd
      : lineStartSec + 4;
  return {
    lineStartSec,
    lineEndSec,
    lineDurSec: Math.max(0.4, lineEndSec - lineStartSec),
  };
}

/**
 * Estimate when the highlighted word is sung inside the LRC line.
 * Uses Unicode code points so accented / non-Latin scripts stay accurate.
 * Line duration is capped so sparse LRC (long instrumental gaps) cannot shove
 * the seek many seconds past the real sung moment.
 */
export function estimateWordSongTimeSec(input: {
  timestamp_ms: number;
  snippet: string;
  char_start: number;
  char_end: number;
  line_end_ms?: number | null;
}): number {
  const { lineStartSec, lineDurSec } = lineBoundsSec(input);
  const effectiveDur = Math.min(lineDurSec, MAX_LINE_DUR_FOR_WORD_EST_SEC);
  const chars = Array.from(String(input.snippet || ''));
  const len = Math.max(1, chars.length);
  // Alignment may be UTF-16 indices; clamp into code-point space.
  const start = Math.max(0, Math.min(Number(input.char_start) || 0, len));
  const end = Math.max(start, Math.min(Number(input.char_end) || start, len));
  const mid = (start + end) / 2;
  const frac = Math.min(1, Math.max(0, mid / len));
  // Bias toward line start: small char fraction on a capped window.
  return lineStartSec + frac * effectiveDur;
}

export function formatPreviewWindowLabel(offsetSec: number, lengthSec = 30): string {
  const start = Math.max(0, Math.floor(offsetSec));
  const end = start + lengthSec;
  const fmt = (s: number) => `${Math.floor(s / 60)}:${String(s % 60).padStart(2, '0')}`;
  return `${fmt(start)}–${fmt(end)}`;
}

/** Deezer: 30–60s for long tracks. iTunes: usually from 0 (or unknown). */
export function resolvePreviewOffsetSec(input: {
  preview_offset?: number | null;
  preview_provider?: string | null;
  duration_seconds?: number | null;
}): number {
  const provider = String(input.preview_provider || 'deezer').toLowerCase();
  if (provider === 'itunes' || provider.startsWith('itunes')) {
    // Apple clips are usually the opening ~30s unless we know otherwise.
    return 0;
  }
  if (typeof input.preview_offset === 'number' && Number.isFinite(input.preview_offset)) {
    return Math.max(0, input.preview_offset);
  }
  const dur = Number(input.duration_seconds) || 0;
  if (dur > 60) return 30;
  if (dur > 30) return dur - 30;
  return 0;
}

export type DeezerHearWindow = {
  seekTo: number;
  stopAt: number;
  inWindow: boolean;
  /** False when the preview bytes cannot contain the lyric — do not play a misleading edge clip. */
  shouldPlay: boolean;
  relative: number;
  wordSongTimeSec: number;
};

/**
 * Map full-song lyric time into the 30s preview element timeline.
 * Word-centered: start ~1s before the word so the sung word matches the phrase.
 */
export function computeDeezerHearWindow(input: {
  timestamp_ms: number;
  snippet: string;
  char_start: number;
  char_end: number;
  preview_offset?: number;
  preview_provider?: string | null;
  duration_seconds?: number | null;
  preview_len?: number;
  line_end_ms?: number | null;
}): DeezerHearWindow {
  const PREVIEW_LEN = input.preview_len ?? 30;
  const WORD_LEAD = 1.0; // seconds before the sung word
  const WORD_TAIL = 5.5;
  const TARGET_CLIP = 8;
  const MIN_CLIP = 5;

  const wordSongTimeSec = estimateWordSongTimeSec(input);
  const offset = resolvePreviewOffsetSec(input);
  const relativeWord = wordSongTimeSec - offset;
  const inWindow = relativeWord >= 0.4 && relativeWord <= PREVIEW_LEN - 0.8;

  if (inWindow) {
    let seekTo = Math.max(0, relativeWord - WORD_LEAD);
    let stopAt = Math.min(PREVIEW_LEN, relativeWord + WORD_TAIL);
    if (stopAt - seekTo < TARGET_CLIP) {
      stopAt = Math.min(PREVIEW_LEN, seekTo + TARGET_CLIP);
    }
    if (stopAt - seekTo < MIN_CLIP) {
      seekTo = Math.max(0, stopAt - MIN_CLIP);
    }
    // Keep the word inside the clip with a small margin.
    if (seekTo > relativeWord - 0.35) {
      seekTo = Math.max(0, relativeWord - WORD_LEAD);
    }
    if (stopAt < relativeWord + 1.5) {
      stopAt = Math.min(PREVIEW_LEN, relativeWord + WORD_TAIL);
    }
    stopAt = Math.max(stopAt, Math.min(PREVIEW_LEN, seekTo + MIN_CLIP));
    return {
      seekTo: Math.round(seekTo * 100) / 100,
      stopAt: Math.round(stopAt * 100) / 100,
      inWindow: true,
      shouldPlay: true,
      relative: relativeWord,
      wordSongTimeSec,
    };
  }

  // Outside the preview cut — do not play a random edge (that lied to users).
  return {
    seekTo: 0,
    stopAt: 0,
    inWindow: false,
    shouldPlay: false,
    relative: relativeWord,
    wordSongTimeSec,
  };
}

/** Spotify full-track seek: start just before the word, play a short context clip. */
export function computeSpotifyHearClip(input: {
  timestamp_ms: number;
  snippet: string;
  char_start: number;
  char_end: number;
  line_end_ms?: number | null;
}): { positionMs: number; stopAfterMs: number; wordSongTimeSec: number } {
  const WORD_LEAD_MS = 1000;
  const WORD_TAIL_MS = 5500;
  const TARGET_MS = 8000;
  const MIN_MS = 5000;
  const MAX_MS = 14000;

  const wordSongTimeSec = estimateWordSongTimeSec(input);
  const wordMs = Math.round(wordSongTimeSec * 1000);
  const positionMs = Math.max(0, wordMs - WORD_LEAD_MS);
  let stopAfterMs = WORD_LEAD_MS + WORD_TAIL_MS;
  if (stopAfterMs < TARGET_MS) stopAfterMs = TARGET_MS;
  if (stopAfterMs < MIN_MS) stopAfterMs = MIN_MS;
  stopAfterMs = Math.min(MAX_MS, stopAfterMs);

  return {
    positionMs,
    stopAfterMs,
    wordSongTimeSec,
  };
}

File: mobile/lib/utils/hear_it_timing.dart
/// Hear-it timing helpers — map lyric lines into Deezer / iTunes preview seek.
/// Mirrors `client/src/lib/hearItTiming.ts` so Flutter and web stay in sync.

const double _maxLineDurForWordEstSec = 5;

class LineBounds {
  const LineBounds({
    required this.lineStartSec,
    required this.lineEndSec,
    required this.lineDurSec,
  });

  final double lineStartSec;
  final double lineEndSec;
  final double lineDurSec;
}

LineBounds lineBoundsSec({
  required num timestampMs,
  num? lineEndMs,
}) {
  final lineStartSec = timestampMs.toDouble() / 1000;
  final start = lineStartSec < 0 ? 0.0 : lineStartSec;
  final rawEndSec = lineEndMs != null ? lineEndMs.toDouble() / 1000 : double.nan;
  final lineEndSec =
      rawEndSec.isFinite && rawEndSec > start + 0.35 ? rawEndSec : start + 4;
  final dur = lineEndSec - start;
  return LineBounds(
    lineStartSec: start,
    lineEndSec: lineEndSec,
    lineDurSec: dur < 0.4 ? 0.4 : dur,
  );
}

/// Estimate when the highlighted word is sung inside the LRC line.
double estimateWordSongTimeSec({
  required num timestampMs,
  required String snippet,
  required num charStart,
  required num charEnd,
  num? lineEndMs,
}) {
  final bounds = lineBoundsSec(timestampMs: timestampMs, lineEndMs: lineEndMs);
  final effectiveDur = bounds.lineDurSec < _maxLineDurForWordEstSec
      ? bounds.lineDurSec
      : _maxLineDurForWordEstSec;
  final chars = snippet.runes.toList();
  final len = chars.isEmpty ? 1 : chars.length;
  var start = charStart.toDouble();
  if (start < 0) start = 0;
  if (start > len) start = len.toDouble();
  var end = charEnd.toDouble();
  if (end < start) end = start;
  if (end > len) end = len.toDouble();
  final mid = (start + end) / 2;
  final frac = (mid / len).clamp(0.0, 1.0);
  return bounds.lineStartSec + frac * effectiveDur;
}

/// Deezer: 30–60s for long tracks. iTunes: usually from 0.
double resolvePreviewOffsetSec({
  num? previewOffset,
  String? previewProvider,
  num? durationSeconds,
}) {
  final provider = (previewProvider ?? 'deezer').toLowerCase();
  if (provider == 'itunes' || provider.startsWith('itunes')) {
    return 0;
  }
  if (previewOffset != null && previewOffset.isFinite) {
    final o = previewOffset.toDouble();
    return o < 0 ? 0 : o;
  }
  final dur = durationSeconds?.toDouble() ?? 0;
  if (dur > 60) return 30;
  if (dur > 30) return dur - 30;
  return 0;
}

class DeezerHearWindow {
  const DeezerHearWindow({
    required this.seekTo,
    required this.stopAt,
    required this.inWindow,
    required this.shouldPlay,
    required this.relative,
    required this.wordSongTimeSec,
  });

  final double seekTo;
  final double stopAt;
  final bool inWindow;
  final bool shouldPlay;
  final double relative;
  final double wordSongTimeSec;
}

double _round2(double v) => (v * 100).round() / 100;

/// Map full-song lyric time into the 30s preview element timeline.
DeezerHearWindow computeDeezerHearWindow({
  required num timestampMs,
  required String snippet,
  required num charStart,
  required num charEnd,
  num? previewOffset,
  String? previewProvider,
  num? durationSeconds,
  double previewLen = 30,
  num? lineEndMs,
}) {
  const wordLead = 1.0;
  const wordTail = 5.5;
  const targetClip = 8.0;
  const minClip = 5.0;

  final wordSongTimeSec = estimateWordSongTimeSec(
    timestampMs: timestampMs,
    snippet: snippet,
    charStart: charStart,
    charEnd: charEnd,
    lineEndMs: lineEndMs,
  );
  final offset = resolvePreviewOffsetSec(
    previewOffset: previewOffset,
    previewProvider: previewProvider,
    durationSeconds: durationSeconds,
  );
  final relativeWord = wordSongTimeSec - offset;
  final inWindow = relativeWord >= 0.4 && relativeWord <= previewLen - 0.8;

  if (!inWindow) {
    return DeezerHearWindow(
      seekTo: 0,
      stopAt: 0,
      inWindow: false,
      shouldPlay: false,
      relative: relativeWord,
      wordSongTimeSec: wordSongTimeSec,
    );
  }

  var seekTo = relativeWord - wordLead;
  if (seekTo < 0) seekTo = 0;
  var stopAt = relativeWord + wordTail;
  if (stopAt > previewLen) stopAt = previewLen;
  if (stopAt - seekTo < targetClip) {
    stopAt = seekTo + targetClip;
    if (stopAt > previewLen) stopAt = previewLen;
  }
  if (stopAt - seekTo < minClip) {
    seekTo = stopAt - minClip;
    if (seekTo < 0) seekTo = 0;
  }
  if (seekTo > relativeWord - 0.35) {
    seekTo = relativeWord - wordLead;
    if (seekTo < 0) seekTo = 0;
  }
  if (stopAt < relativeWord + 1.5) {
    stopAt = relativeWord + wordTail;
    if (stopAt > previewLen) stopAt = previewLen;
  }
  final minStop = seekTo + minClip;
  if (stopAt < minStop) {
    stopAt = minStop > previewLen ? previewLen : minStop;
  }

  return DeezerHearWindow(
    seekTo: _round2(seekTo),
    stopAt: _round2(stopAt),
    inWindow: true,
    shouldPlay: true,
    relative: relativeWord,
    wordSongTimeSec: wordSongTimeSec,
  );
}
